dnf5/commands/list/list.cpp

```diff
@@ -0,0 +1,284 @@
+
+    autoremove = std::make_unique<libdnf::cli::session::BoolOption>(
+        *this, "autoremove", '\0', "List packages which will be removed by the 'dnf autoremove' command.", false);
+    conflicts->push_back(autoremove->arg);
+
+    installed->arg->set_conflict_arguments(conflicts);
+    available->arg->set_conflict_arguments(conflicts);
+    extras->arg->set_conflict_arguments(conflicts);
+    obsoletes->arg->set_conflict_arguments(conflicts);
+    recent->arg->set_conflict_arguments(conflicts);
+    upgrades->arg->set_conflict_arguments(conflicts);
+    autoremove->arg->set_conflict_arguments(conflicts);
+}
+
+void ListCommand::configure() {
+    // TODO(mblaha): do not force expired metadata sync if not explicitely required
+    pkg_narrow = PkgNarrow::ALL;
+    Context::LoadAvailableRepos load_available = Context::LoadAvailableRepos::ENABLED;
+    bool load_system = true;
+    if (available->get_value()) {
+        pkg_narrow = PkgNarrow::AVAILABLE;
+    } else if (installed->get_value()) {
+        load_available = Context::LoadAvailableRepos::NONE;
+        pkg_narrow = PkgNarrow::INSTALLED;
+    } else if (extras->get_value()) {
+        pkg_narrow = PkgNarrow::EXTRAS;
+    } else if (obsoletes->get_value()) {
+        pkg_narrow = PkgNarrow::OBSOLETES;
+    } else if (recent->get_value()) {
+        load_system = false;
+        pkg_narrow = PkgNarrow::RECENT;
+    } else if (upgrades->get_value()) {
+        pkg_narrow = PkgNarrow::UPGRADES;
+    } else if (autoremove->get_value()) {
+        load_available = Context::LoadAvailableRepos::NONE;
+        pkg_narrow = PkgNarrow::AUTOREMOVE;
+    }
+    auto & context = get_context();
+    context.set_load_available_repos(load_available);
+    context.set_load_system_repo(load_system);
+}
+
+std::unique_ptr<libdnf::cli::output::PackageListSections> ListCommand::create_output() {
+    auto out = std::make_unique<libdnf::cli::output::PackageListSections>();
+    out->setup_cols();
+    return out;
+}
+
+void ListCommand::run() {
+    auto & ctx = get_context();
+    auto & config = ctx.base.get_config();
+    auto showduplicates = config.showdupesfromrepos().get_value();
+
+    libdnf::rpm::PackageQuery full_package_query(ctx.base);
+    libdnf::rpm::PackageQuery base_query(ctx.base);
+
+    // pre-select by patterns
+    if (!pkg_specs.empty()) {
+        base_query = libdnf::rpm::PackageQuery(ctx.base, libdnf::sack::ExcludeFlags::APPLY_EXCLUDES, true);
+        libdnf::ResolveSpecSettings settings{.with_nevra = true, .with_provides = false, .with_filenames = false};
+        for (const auto & spec : pkg_specs) {
+            libdnf::rpm::PackageQuery pkg_query(full_package_query);
+            pkg_query.resolve_pkg_spec(spec, settings, true);
+            base_query |= pkg_query;
+        }
+    }
+
+    // did any package match the criteria?
+    bool package_matched = false;
+    // output table
+    auto sections = create_output();
+
+    switch (pkg_narrow) {
```

Comment:
Do we want to expose list and info functionality for DBUS users? Or queries are enough?

Reply:
Only note: In DNF4 the code was shared with `repository-pkgs list` command, but in DNF5 we will provide the functionality using a different way.

Reply:
> Do we want to expose list and info functionality for DBUS users? Or queries are enough?

Actually it's currently the other way round. We have `list` functionality on dbus (or something very similar), but we do not support queries there. The problem is that list functionality is "state-less" and always work on top of all packages (or a set limited by pkg_narrow). To implement queries we would need to store query state on the server side - to be able to gradually implement filters.